README.md
<h1>Todo App</h1>
<p>
    This is a simple JavaScript based Todo app that allows users to add, filter,
    and clear tasks. The app uses the DOM to manipulate the HTML and
    localStorage to save the tasks.
</p>
<h2>Getting Started</h2>
<p>
    To get started, simply clone or download the repository and open the
    <code>index.html</code> file in your browser. You should be able to see the
    Todo app interface.
</p>
<h2>Features</h2>
<ul>
    <li>
        Add tasks by typing into the input field and clicking the "Add" button
    </li>
    <li>
        Filter tasks by clicking the "All", "Active", or "Completed" buttons
    </li>
    <li>Clear all tasks by clicking the "Clear" button</li>
    <li>
        Tasks are saved to localStorage, so they will persist even after the
        browser is closed or refreshed
    </li>
</ul>
<h2>Code Structure</h2>
<p>
    The app uses several variables such as <code>addBtn</code>,
    <code>todoCounter</code>, <code>taskTasks</code>,
    <code>pendingFilter</code>, <code>clearAll</code>, <code>todoInput</code>,
    <code>todoList</code>, and <code>localTodo</code> to select different parts
    of the HTML and manipulate them.
</p>
<p>
    The app uses an array <code>todoSaves</code> to store the tasks and another
    array <code>filteredTodoSaves</code> to store the filtered tasks. The
    <code>filterTodoSavesFunc()</code> function is used to filter the tasks
    based on the filter mode (<code>all</code>, <code>active</code>, or
    <code>completed</code>) and the <code>updateHTML(addNewTodo)</code> function
    is used to update the DOM and refresh the list of tasks.
</p>
<h2>Built With</h2>
<ul>
    <li>HTML</li>
    <li>SCSS</li>
    <li>CSS</li>
    <li>JavaScript</li>
    <li>KnockoutJS</li>
</ul>
<h2>Author</h2>
<ul>
    <li><a href="https://github.com/Ali-Sdg90" target="_new">Ali-Sdg90</a></li>
</ul>
<h2>Steps</h2>
<ul>
    <li>
        <p>
            Todo Step 3 :
            <a
                href="https://github.com/Ali-Sdg90/Todo-App/tree/2471d1bfbd34de78b451684a90d3c6db4ac20350"
                target="_new"
                >GitHub</a
            >
            |
            <a href="https://jsfiddle.net/Ali_Sdg90/v60okbum/" target="_new"
                >Jsfiddle</a
            >
        </p>
    </li>
    <li>
        <p>
            Todo Step 4 :
            <a
                href="https://github.com/Ali-Sdg90/Todo-App/tree/1ba75aa5ab8cb51396f7a46678a6b9791e9a0f14"
                target="_new"
                >GitHub</a
            >
            |
            <a href="https://jsfiddle.net/Ali_Sdg90/k4Lwxare/" target="_new"
                >Jsfiddle</a
            >
        </p>
    </li>
    <li>
        <p>
            Todo Step 5 :
            <a
                href="https://github.com/Ali-Sdg90/Todo-App/tree/543a742de6327389adffd28ab8a93fe58f8ec82e"
                target="_new"
                >GitHub</a
            >
            |
            <a href="https://jsfiddle.net/Ali_Sdg90/v7nac15g/4/" target="_new"
                >Jsfiddle</a
            >
        </p>
    </li>
</ul>


History
STEP 3 :
This program was written on November 24, 2021, and I planned a writing marathon challenge for myself.
I wrote in 4 hours from 6 pm to 10 pm, and the next day I made some small corrections.
It is the first program that I recorded completely and time-lapsed :)
Time lapse link : https://youtu.be/XX-UKi4Pwr0

#AZADI
---------------------------------------------

STEP 4:
Step 4 code started during my university semester 4 exams and I spent 8 days on it from Jan 17, 2023 to Jan 25, 2023. During this time, 
I got to know ChatGPT and I got help from it for the README, it's really awesome!
The situation in Iran is awful, and a proposal was approved to add the IRGC to the list of terrorists in Europe...
Life is really hard here, I hope you are in a better situation when you are reading this...

#AZADI
---------------------------------------------

STEP 5:
We should have added knockout.js and MVVM to the todo app. Knockout.js is very interesting, but you have to approach it differently to do anything. From one perspective, updating the site becomes easy, but coding it is hard, and it can really melt your brain...
The situation in Iran is worse than before. The dollar has hit an unprecedented record of 60 tomans, and every day there is news of high school girls being poisoned... It's very scary, very, very scary.
Two days ago, finally after 231 days, I had my CS interview. I hope everything goes well...
Last time, I wished that when you read this next time, things would be better. It didn't happen for me, but I hope it happened for you :)

#AZADI


knockout.js/js/script.js
const filterBackground = document.getElementsByClassName("filter-background");
const filtertext = document.getElementsByClassName("filter-text");
const filterDivBackground = document.getElementsByClassName(
    "filter-div-background"
);

// style for nav btns
for (let i = 0; i < 3; i++) {
    filtertext[i].addEventListener("click", function () {
        for (let j = 0; j < 3; j++) {
            filterBackground[j].style.fill = "rgb(198, 198, 198)";
            filterDivBackground[j].style.zIndex = "1";
        }
        filterBackground[i].style.fill = "white";
        filterDivBackground[i].style.zIndex = "2";
    });
}

// Prevent the browser's default behavior of refreshing the page when 'enter' is
// pressed while an input is in focus :
document.querySelector("form").addEventListener("submit", function (event) {
    event.preventDefault();
});

// Simulate clicking on the addBtn by pressing the Enter key.
document.addEventListener("keyup", function (event) {
    if (event.key === "Enter") {
        document.getElementById("add-btn").dispatchEvent(new Event("click"));
    }
});


knockout.js/js/knockout.js
const todoInput = document.getElementById("todo-input");

// View Model
const VM = {
    // All saved todos
    todos: ko.observableArray([]),

    // Input value
    todoText: ko.observable(""),

    // For edit todo
    acceptInput: ko.observable(true),

    addTodo: function () {
        if (this.acceptInput() && todoInput.value) {
            VM.todos.push(new addNewTodo(VM.todoText()));
            VM.todoText("");
        }
    },
    todosFilter: ko.observable("all"), // all - active - completed

    pending: function () {
        if (this.todosFilter() === "all") {
            return "";
        } else {
            return this.todosFilter();
        }
    },
    todosFilterAll: function () {
        console.log("all");
        this.todosFilter("all");
    },
    todosFilterActive: function () {
        console.log("active");
        this.todosFilter("active");
    },
    todosFilterCompleted: function () {
        console.log("completed");
        this.todosFilter("completed");
    },

    clearAll: function () {
        this.todos(
            this.todos().filter(function (todo) {
                return !VM.showTodos().includes(todo);
            })
        );
    },

    // Only if showTodos is empty
    addBr: function () {
        if (!VM.showTodos().length) {
            return "<br>";
        }
    },
};

// Obj constructor for addTodo
function addNewTodo(todoText) {
    this.todoText = todoText;
    this.isActive = ko.observable(true);
    this.editTodo = function () {
        todoInput.value = this.todoText;
        VM.acceptInput(false);
        document.getElementById("add-btn").addEventListener(
            "click",
            function () {
                this.todoText = todoInput.value;
                VM.acceptInput(true);
                console.log(this.todoText, todoInput.value);
            },
            { once: true }
        );
    };
    this.changeCompleteState = function () {
        this.isActive(!this.isActive());
    };
    this.deleteTodo = function () {
        VM.todos.remove(this);
    };

    // Log changes in todo's isActive
    this.isActive.subscribe(function (value) {
        console.log(todoText, "=>", value);
    });
}

VM.showTodos = ko.computed(function () {
    let filteredList = function () {
        switch (VM.todosFilter()) {
            case "active":
                return VM.todos().filter(function (todo) {
                    return todo.isActive();
                });
            case "completed":
                return VM.todos().filter(function (todo) {
                    return !todo.isActive();
                });
            default:
                return VM.todos();
        }
    };

    // Nav position
    if (filteredList().length < 7) {
        let filterTop = 35.5 - filteredList().length * 3 + "vh";
        document.getElementById("filter-nav-background").style.top = filterTop;
        document.getElementById("filter-nav-text").style.top = filterTop;
    } else {
        document.getElementById("filter-nav-background").style.top = "14.85vh";
        document.getElementById("filter-nav-text").style.top = "14.85vh";
    }
    console.log(filteredList());

    // New todo placed at the top of the list
    return filteredList().slice(0).reverse();
});

// remaining task text
taskTasks = ko.computed(function () {
    let output = "";
    if (VM.showTodos().length == 1) {
        output += "task";
    } else {
        output += "tasks";
    }
    if (VM.todosFilter() === "all") {
        output += " in total";
    }
    return output;
});

ko.applyBindings(VM);


js/updateHTML.js
import {
    filterTodoSavesFunc,
    editBtn,
    completeBtn,
    deleteBtn,
} from "./script.js";
import {
    localTodo,
    todoSaves,
    filteredTodoSaves,
    editTodo,
    firstTopFilter,
    filterList,
} from "./var-js.js";
import {
    addBtn,
    todoCounter,
    taskTasks,
    pendingFilter,
    clearAll,
    todoInput,
    todoList,
} from "./var-dom.js";

// Update the DOM
// Create new tasks with buttons and addEventListeners to them
// Set the position of navigation filters
// Update pending tasks in different modes :
function updateHTML(addNewTodo) {
    todoInput.value = "";
    filterTodoSavesFunc();
    if (!filteredTodoSaves.length) todoList.innerHTML = "<br />";
    else todoList.innerHTML = "";
    for (let i = filteredTodoSaves.length; i > 0; ) {
        i--;
        todoList.innerHTML += `
        <span id="todo-number${i}">
            <div id="work-number${i}">${filteredTodoSaves[i].todo}</div>
            <div id="edit-number${i}">
            <svg xmlns="http://www.w3.org/2000/svg"  viewBox="0 0 30 30" width="30px" height="30px"><path d="M24 11l2.414-2.414c.781-.781.781-2.047 0-2.828l-2.172-2.172c-.781-.781-2.047-.781-2.828 0L19 6 24 11zM17 8L5.26 19.74C7.886 19.427 6.03 21.933 7 23c.854.939 3.529-.732 3.26 1.74L22 13 17 8zM4.328 26.944l-.015-.007c-.605.214-1.527-.265-1.25-1.25l-.007-.015L4 23l3 3L4.328 26.944z"/></svg>
            </div>
            <div id="complete-number${i}">
            <svg xmlns="http://www.w3.org/2000/svg"  viewBox="0 0 24 24" width="24px" height="24px"><path d="M 5 3 C 3.897 3 3 3.897 3 5 L 3 19 C 3 20.103 3.897 21 5 21 L 19 21 C 20.103 21 21 20.103 21 19 L 21 9.2421875 L 19 11.242188 L 19.001953 19 L 5 19 L 5 5 L 16.757812 5 L 18.757812 3 L 5 3 z M 21.292969 3.2929688 L 11 13.585938 L 7.7070312 10.292969 L 6.2929688 11.707031 L 11 16.414062 L 22.707031 4.7070312 L 21.292969 3.2929688 z"/></svg>
            </div>
            <div id="delete-number${i}">
            <svg xmlns="http://www.w3.org/2000/svg"  viewBox="0 0 48 48" width="48px" height="48px"><path fill="none" stroke="#000" stroke-miterlimit="10" stroke-width="3" d="M29.5,11.5V11c0-3-2.5-5.5-5.5-5.5S18.5,8,18.5,11v0.5"/><line x1="7.5" x2="40.5" y1="11.5" y2="11.5" fill="none" stroke="#000" stroke-linecap="round" stroke-miterlimit="10" stroke-width="3"/><line x1="36.5" x2="38" y1="27" y2="11.5" fill="none" stroke="#000" stroke-linecap="round" stroke-miterlimit="10" stroke-width="3"/><path fill="none" stroke="#000" stroke-linecap="round" stroke-miterlimit="10" stroke-width="3" d="M10.7,18.6l2,20.3c0.2,2.1,1.9,3.6,4,3.6h14.7c2.1,0,3.8-1.6,4-3.6l0.5-4.8"/></svg>
            </div>
        </span>
        `;
    }
    for (let i = 0; i < filteredTodoSaves.length; i++) {
        document
            .getElementById(`edit-number${i}`)
            .addEventListener("click", function () {
                editBtn(i);
            });
        document
            .getElementById(`complete-number${i}`)
            .addEventListener("click", function () {
                completeBtn(i);
            });
        document
            .getElementById(`delete-number${i}`)
            .addEventListener("click", function () {
                deleteBtn(i);
            });
        if (filteredTodoSaves[i].isComplete) {
            document
                .getElementById(`work-number${i}`)
                .classList.toggle("complete-todo");
        }
    }
    if (addNewTodo) {
        const newTodo = document.getElementById(
            `todo-number${filteredTodoSaves.length - 1}`
        );
        newTodo.style.transform = "translate(0, -3.5vh)";
        newTodo.style.opacity = "0";
        setTimeout(() => {
            newTodo.style.opacity = "1";
            newTodo.style.transform = "translate(0, 0)";
        }, 10);
    }
    todoCounter.textContent = filteredTodoSaves.length;
    if (filteredTodoSaves.length == 1) {
        taskTasks.textContent = "task";
    } else {
        taskTasks.textContent = "tasks";
    }
    switch (filterList) {
        case "all":
            pendingFilter.textContent = "";
            if (filteredTodoSaves.length == 1) {
                taskTasks.textContent = "task in total";
            } else {
                taskTasks.textContent = "tasks in total";
            }
            break;
        case "active":
            pendingFilter.textContent = "active";
            break;
        case "completed":
            pendingFilter.textContent = "completed";
            break;
    }
    if (filteredTodoSaves.length < 8) {
        let filterTop = firstTopFilter - filteredTodoSaves.length * 2.95 + "vh";
        document.getElementById("filter-nav-background").style.top = filterTop;
        document.getElementById("filter-nav-text").style.top = filterTop;
    } else {
        document.getElementById("filter-nav-background").style.top = "14.85vh";
        document.getElementById("filter-nav-text").style.top = "14.85vh";
    }
}

export { updateHTML };
